src/api/article/article.model.js

```diff
@@ -46,119 +44,88 @@ const joiArticleSchema = Joi.object({
-  const { error } = Joi.validate(this.images, IMAGES_SCHEMA[this.type].required());
-  if (error !== null) {
-    const errors = {};
-    error.details.forEach(({ path, type }) => {
-      set(errors, ['images', ...path], type);
-    });
+const getImagesSchema = type =>
+  Joi.object(
+    IMAGES_BY_TYPE[type].reduce((acc, cur) => {
+      acc[cur] = Joi.image();
+      return acc;
+    }, {})
+  );
+
+export const validateArticle = data => {
+  const { type } = data;
+  const schema = joiArticleSchema.keys({
+    images: getImagesSchema(type).required(),
+    // TODO: conditional require `video` or `audio`
+  });
+  return defaultValidator(data, schema);
+};
 
-    next(new ValidationError(errors));
-  }
-  next();
-});
+const ArticleSchema = joiToMongoose(joiArticleSchema, { usePushEach: true }, validateArticle);
+
+const formatArticle = article =>
+  article
+    ? {
+        ...omit(article.toObject(), [
+          '__v',
+          'collectionId',
+          'video._id',
+          'metadata._id',
+          'metadata.createdBy._id',
+          'metadata.createdBy.displayName',
+          'metadata.updatedBy._id',
+          'metadata.updatedBy.displayName',
+        ]),
+        locales: keyBy(article.locales, 'locale'),
+      }
+    : null;
 
 // includeCollection flag here is to be able to avoid including collections
-// in case we're serializing an article into the ArticleCollection object..
+// in case we're serializing an article into the ArticleCollection object.
 export const serializeArticle = (article, { includeCollection = true } = {}) => {
-  const collectionNavigation = {};
-
-  if (includeCollection && article.collectionId) {
-    collectionNavigation.prev = null;
-    collectionNavigation.next = null;
-
-    const { articles } = article.collectionId;
-    const idx = mapIds(articles).indexOf(getId(article));
-
-    if (idx > 0) {
-      const a = articles[idx - 1].toObject();
-      collectionNavigation.prev = {
-        ...omit(a, ['locales']),
-        locales: keyBy(a.locales, 'locale'),
-      };
-    }
-
-    if (idx !== articles.length - 1) {
-      const a = articles[idx + 1].toObject();
-      collectionNavigation.next = {
-        ...omit(a, ['locales']),
-        locales: keyBy(a.locales, 'locale'),
-      };
-    }
-  }
+  const result = formatArticle(article);
 
-  const result = {
-    ...omit(article.toObject(), [
-      '__v',
-      'collectionId',
-      'video._id',
-      'metadata._id',
-      'metadata.createdBy._id',
-      'metadata.createdBy.displayName',
-      'metadata.updatedBy._id',
-      'metadata.updatedBy.displayName',
-    ]),
-    locales: keyBy(article.locales, 'locale'),
-  };
-
-  if (includeCollection) {
-    result.collection = article.collectionId && {
-      ...omit(article.collectionId.toObject(), ['articles']),
-      ...collectionNavigation,
-    };
+  if (!includeCollection || !article.collectionId) {
+    return result;
   }
 
+  const { articles } = article.collectionId;
```

Comment:
I am not sure I get it. Is `collectionId` somehow pre-populated with the whole collection of articles?

Reply:
yep, looks weird, I agree.